### PR TITLE
Add EvictOldWatchdogViolations().

### PR DESCRIPTION
**Pull request by briantting**
EvictOldWatchdogViolations() evicts unfetched Watchdog violations older than 24 hours. It is called only after GetWatchdogViolations() is called. It is reponsible for ensuring that unfetched Watchdog violations do not accumulate.

b/296308743

Change-Id: Ic225d80382697341f6338d554941d78930e62035

### CODE REVIEW

File: cobalt/watchdog/watchdog.cc [hunk 1]
```diff
@@ -541,35 +544,32 @@ bool Watchdog::Ping(const std::string& name, const std::string& info) {
...
     if (clients.empty()) {
-      // Removes all Watchdog violations.
-      base::JSONWriter::Write(*violations_map_, &watchdog_json_fetched);
+      base::JSONWriter::Write(*violations_map_, &fetched_violations_json);
       if (clear) {
         static_cast<base::DictionaryValue*>(violations_map_.get())->Clear();
         violations_count_ = 0;
         starboard::SbFileDeleteRecursive(GetWatchdogFilePath().c_str(), true);
       }
+      // Gets all Watchdog violations of the given clients.
```
Comment (sherryzy) [inline cobalt/watchdog/watchdog.cc:564]: If clear is false, we should also call EvictOldWatchdogViolations()?

File: cobalt/watchdog/watchdog.cc [hunk 2]
```diff
@@ -579,23 +579,44 @@ std::string Watchdog::GetWatchdogViolations(
...
+    SB_LOG(INFO) << "[Watchdog] Reading violations:\n"
+                 << fetched_violations_json;
   } else {
     SB_LOG(INFO) << "[Watchdog] No violations.";
   }
-  return watchdog_json_fetched;
+  return fetched_violations_json;
+}
+
+void Watchdog::EvictOldWatchdogViolations() {
```
Comment (sherryzy) [inline cobalt/watchdog/watchdog.cc:603]: We should add a test for this function as well.

File: cobalt/watchdog/watchdog.cc (hunk 1 above)
Comment (briantting) [reply cobalt/watchdog/watchdog.cc:564]: agreed, if clear is false we shouldn't touch the json file (including evicting old violations).

File: cobalt/watchdog/watchdog.cc [hunk 3]
```diff
@@ -541,35 +544,32 @@ bool Watchdog::Ping(const std::string& name, const std::string& info) {
 
 std::string Watchdog::GetWatchdogViolations(
     const std::vector<std::string>& clients, bool clear) {
```
Comment (joeltine) [inline cobalt/watchdog/watchdog.cc:546]: Side comment: When is clear ever false? I don't see any real uses of it. Why do we need the "clear" argument at all? @sherryzy

File: cobalt/watchdog/watchdog.cc [hunk 4]
```diff
@@ -579,23 +579,46 @@ std::string Watchdog::GetWatchdogViolations(
...
+                 << fetched_violations_json;
   } else {
     SB_LOG(INFO) << "[Watchdog] No violations.";
   }
-  return watchdog_json_fetched;
+  return fetched_violations_json;
+}
+
+void Watchdog::EvictOldWatchdogViolations() {
+  int64_t current_timestamp = SbTimeToPosix(SbTimeGetNow()) / 1000;
```
Comment (joeltine) [inline cobalt/watchdog/watchdog.cc:606]: Nit: I always find it much easier to read variable names that have the units attached, for example: `current_timestamp_millis`. That way, I'm not guessing what units I'm dealing with here.

File: cobalt/watchdog/watchdog.cc [hunk 5]
```diff
@@ -579,23 +579,46 @@ std::string Watchdog::GetWatchdogViolations(
             violations_count_ -= violations_count;
             if (!static_cast<base::DictionaryValue*>(violations_map_.get())
                      ->empty()) {
-              WriteWatchdogViolations();
+              pending_write_ = true;
```
Comment (joeltine) [inline cobalt/watchdog/watchdog.cc:582]: Why the change here?

File: cobalt/watchdog/watchdog.cc [hunk 6]
```diff
@@ -579,23 +579,46 @@ std::string Watchdog::GetWatchdogViolations(
...
+    base::Value& violation_dict = it.second;
+    base::Value* violations = violation_dict.FindKey("violations");
+    for (auto i = violations->GetList().begin();
+         i != violations->GetList().end(); i++) {
+      int64_t timestamp =
+          std::stoll(i->FindKey("timestampViolationMilliseconds")->GetString());
+      if (timestamp < cutoff_timestamp) {
+        violations->GetList().erase(i);
+        violations_count_--;
+        pending_write_ = true;
```
Comment (joeltine) [inline cobalt/watchdog/watchdog.cc:618]: Does this asynchronous write mean it's possible we delete some violations and they don't actually get deleted? It'll still eventually get deleted if that happens though, right?

**Review by joeltine — CHANGES_REQUESTED** [review]

File: cobalt/watchdog/watchdog.cc (hunk 3 above)
Comment (sherryzy) [reply cobalt/watchdog/watchdog.cc:546]: Well, this should be a @briantting 's question. I don't think we can set this clear flag from the outside Cobalt. But in the cobalt code, we can definitely have the freedom to set the flag.
So either we clean this flag up, or we should handle the case when `clear` is false to be consistent.

Comment (briantting) [reply cobalt/watchdog/watchdog.cc:546]: clear is false occasionally in watchdog_test, the clear flag is never exposed / available to kabuki though

File: cobalt/watchdog/watchdog.cc (hunk 5 above)
Comment (briantting) [reply cobalt/watchdog/watchdog.cc:582]: We don't want to write every single time there's a change. This can cause I/O limitations on certain platforms. Especially here when it could possibly write once per fetched client (100 fetched clients is an additional 100 writes). It will write when it can (given the backoff) in the monitor loop or when Watchdog / Cobalt is terminated or when it's absolutely necessary. Watchdog otherwise operates with the in-memory data structures.

File: cobalt/watchdog/watchdog.cc (hunk 6 above)
Comment (briantting) [reply cobalt/watchdog/watchdog.cc:618]: One way to think about it is the in-memory data structure is the source of truth while Watchdog is running and the on-disk json is always lagging behind and mirroring the in-memory data structure (this is to save on I/O). The on-disk json is synced to the in-memory data structure periodically and always before the on-disk json is accessed. The on-disk json is also synced before Watchdog is shutdown or triggers a crash. There is always a risk that something catastrophic occurs that causes us to lose a violation but that will be true even if we attempt write ASAP. I believe there was some discussion also about how I/O operations (especially when done very often) often poses the greatest risk to causing problems themselves. Pesky hardware.

File: cobalt/watchdog/watchdog.cc (hunk 3 above)
Comment (joeltine) [reply cobalt/watchdog/watchdog.cc:546]: Should we delete it? No one is using it. It's also dangerous to use if those violations don't get fetched. Separate cl maybe?

File: cobalt/watchdog/watchdog.cc (hunk 4 above)
Comment (briantting) [reply cobalt/watchdog/watchdog.cc:606]: done

**Comment by sherryzy** [comment]
The test failures around GetPartialViolationsByClients() in watchdog_test seem related, though I'm not sure how that happens. Can you take a look and fix that?

File: cobalt/watchdog/watchdog.cc (hunk 2 above)
Comment (briantting) [reply cobalt/watchdog/watchdog.cc:599]: added a test. also significantly trimmed GetPartialViolationsByClients test. this was because that test was failing due to testing beyond its scope. it was testing for file writes but that behavior has now changed from always forced to next available monitor loop or shutdown (those 2 write cases are already being tested by other more specific tests)

File: cobalt/watchdog/watchdog.cc (hunk 3 above)
Comment (briantting) [reply cobalt/watchdog/watchdog.cc:546]: separate cl, for now were keeping the clear in this cl

**Review by sherryzy — APPROVED** [review]

**Review by joeltine — APPROVED** [review]
